### PR TITLE
[firebase_remote_config] Remote config/double instance error

### PR DESCRIPTION
**Pull request by IchordeDionysos**
## Description

This pull request addresses an issue with a test where two instances were created shortly after another leading to a `Future already completed` `StateError`.
This was caused due to side-effects of the tests where all tests passed when they were executed at once and the `doubleInstance` test failed when executed alone.

Also in this Pull Request is a fix for the `doubleInstance` error. Realized with a try/catch for only this exact error. I've used this fix because to me it feels much more elegant compared to a fix where I would basically would have to check for the same shortly after.

Alternative solution:

```dart
if (!instanceCompleter.isCompleted) {
  final _remoteConfigInstance = await _getRemoteConfigInstance();
  if (!instanceCompleter.isCompleted) {
    instanceCompleter.complete(_remoteConfigInstance);
  }
}
```

## Related Issues

#195
flutter/flutter#38060
#28

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] If the pull request affects only one plugin, the PR title starts with the name of the plugin in brackets (e.g. [cloud_firestore])
- [x] My PR includes unit or integration tests for *all* changed/updated/fixed behaviors (See [Contributor Guide]).
- [x] All existing and new tests are passing.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] The analyzer (`flutter analyze`) does not report any problems on my PR.
- [x] I read and followed the [Flutter Style Guide].
- [x] I updated pubspec.yaml with an appropriate new version according to the [pub versioning philosophy].
- [x] I updated CHANGELOG.md to add a description of the change.
- [x] I signed the [CLA].
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require plugin users to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (please indicate a breaking change in CHANGELOG.md and increment major revision).
- [x] No, this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/FirebaseExtended/flutterfire/blob/master/CONTRIBUTING.md
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[pub versioning philosophy]: https://www.dartlang.org/tools/pub/versioning
[CLA]: https://cla.developers.google.com/

### CODE REVIEW

**Comment by IchordeDionysos** [comment]
Is the failing check known to behave flakily? 🤔 
I can't figure out what code changes the check failure would have triggered.
Also, it says `fatal: repository 'https://cdn.cocoapods.org/' not found`...

**Comment by IchordeDionysos** [comment]
Also, I'm open to alternative solutions to fix the error "Future already completed", 

I just wanted to fix the tests 
and get the bug fix going, 
because when I attach a debugger it always stops first at the exception/error and not my breakpoint, which is annoying.

Interestingly, this does work and gives back a `RemoteConfig` instance it's just that the error still occurs.

File: packages/firebase_remote_config/lib/src/remote_config.dart [hunk 1]
```diff
@@ -25,17 +25,24 @@ class RemoteConfig extends ChangeNotifier {
...
 
   /// Gets the instance of RemoteConfig for the default Firebase app.
   static Future<RemoteConfig> get instance async {
-    if (!_instanceCompleter.isCompleted) {
-      _instanceCompleter.complete(await _getRemoteConfigInstance());
+    if (!instanceCompleter.isCompleted) {
+      try {
+        instanceCompleter.complete(await _getRemoteConfigInstance());
+      } on StateError catch (error) {
+        if (error.message != 'Future already completed') rethrow;
```
Comment (collinjackson) [inline packages/firebase_remote_config/lib/src/remote_config.dart]: It seems like we could just check if the `instanceCompleter` is complete at this point, instead of checking the exception in a fragile way.

File: packages/firebase_remote_config/lib/src/remote_config.dart [hunk 2]
```diff
@@ -25,17 +25,24 @@ class RemoteConfig extends ChangeNotifier {
   RemoteConfigSettings _remoteConfigSettings;
 
   DateTime get lastFetchTime => _lastFetchTime;
+
   LastFetchStatus get lastFetchStatus => _lastFetchStatus;
+
   RemoteConfigSettings get remoteConfigSettings => _remoteConfigSettings;
 
-  static Completer<RemoteConfig> _instanceCompleter = Completer<RemoteConfig>();
+  @visibleForTesting
```
Comment (collinjackson) [inline packages/firebase_remote_config/lib/src/remote_config.dart]: We're generally trying to avoid @visibleForTesting APIs since they clutter the public logs, though I don't see another solution at this point.

File: packages/firebase_remote_config/lib/src/remote_config.dart (hunk 1 above)
Comment (IchordeDionysos) [reply packages/firebase_remote_config/lib/src/remote_config.dart]: I have rewritten the code to use an if statement.

File: packages/firebase_remote_config/lib/src/remote_config.dart (hunk 2 above)
Comment (IchordeDionysos) [reply packages/firebase_remote_config/lib/src/remote_config.dart]: Totally agree, I tried to find an alternative first before using it, but have to change it back somehow...

**Comment by IchordeDionysos** [comment]
@collinjackson @kroikie Any updates on this?
Would it be okay to use @visibleForTesting in this case?

File: packages/firebase_remote_config/lib/src/remote_config.dart (hunk 2 above)
Comment (IchordeDionysos) [reply packages/firebase_remote_config/lib/src/remote_config.dart:33]: @FirebaseExtended/invertase @collinjackson is there any other solution to be able to test for this edge case?

**Comment by IchordeDionysos** [comment]
@FirebaseExtended/invertase @Ehesp

It would be cool to have this fix landed. I've heard a couple of times in our company alone, that people spent quite some time getting to fix this issue, just because it sends a false-positive.

This adds a missing test case, where Remote Config is initialized twice in parallel, which leads to an error.

The current status of this PR is that @collinjackson wasn't too happy about using `@visibleForTesting`, but we found no other way to do so.
To fix the now failing test, I made some adjustments that there's no `Future already completed` error anymore.

I'll be happy to update the CHANGELOG.md if you are happy with the current implementation of the tests.

**Comment by IchordeDionysos** [comment]
Closing in favor of #4186 which makes this PR obsolete!
